src/Switch.js

```diff
@@ -1,71 +1,51 @@
-import React, { Component, PropTypes } from 'react';
-import DefaultRenderer from './DefaultRenderer';
+import React, { PropTypes } from 'react';
+import TabBar from './TabBar';
 import Actions from './Actions';
 
-export default class Switch extends Component {
+export default function Switch(props) {
+  const navState = props.navigationState;
 
-  static propTypes = {
-    navigationState: PropTypes.object,
-    onNavigate: PropTypes.func,
-    selector: PropTypes.func,
-    unmountScenes: PropTypes.bool,
-  };
+  const selector = props.selector;
+  if (!selector) console.error('Selector should be defined.');
 
-  constructor(props, context) {
-    super(props, context);
-    this.state = {};
-    this.updateState = this.updateState.bind(this);
-  }
-
-  componentDidMount() {
-    this.updateState(this.props);
-  }
+  const selectedKey = selector(props);
+  if (!selectedKey) console.error('Selector should return key.');
 
-  componentWillReceiveProps(nextProps) {
-    this.updateState(nextProps);
-  }
+  const selected = navState.children.filter(el => el.sceneKey === selectedKey);
+  if (!selected) console.error(`A scene for key “${selectedKey}” does not exist.`);
 
-  shouldComponentUpdate(nextProps, nextState) {
-    if (this.state.navigationState !== nextState.navigationState) {
-      return true;
+  let index = -1;
+  navState.children.forEach((el, i) => {
+    if (el.sceneKey === selectedKey) {
+      index = i;
     }
-    return false;
-  }
-
-  updateState(props) {
-    const navState = props.navigationState;
-
-    const selector = props.selector;
-    if (!selector) console.error('Selector should be defined.');
+  });
+  if (index === -1) console.error(`A scene for key “${selectedKey}” does not exist.`);
 
-    const selectedKey = selector(props);
-    if (!selectedKey) console.error('Selector should return key.');
-
-    const selected = navState.children.filter(el => el.sceneKey === selectedKey);
-    if (!selected) console.error(`A scene for key “${selectedKey}” does not exist.`);
-
-    const navigationState = selected[0];
-    if (!navigationState) console.error(`Cannot find a scene with key “${selectedKey}”`);
-
-    if (navigationState.sceneKey !== navState.children[navState.index].sceneKey) {
-      if (props.unmountScenes) {
+  let navigationState = index !== navState.index ? { ...navState, index } : navState;
+  if (index !== navState.index) {
+    if (props.unmountScenes) {
+      setTimeout(() => {
         Actions[selectedKey]({ unmountScenes: true });
-      } else {
+      }, 10);
+    } else {
+      setTimeout(() => {
         Actions[selectedKey]();
-      }
+      }, 10);
     }
-    this.setState({ navigationState });
-  }
-
-  render() {
-    if (this.state.navigationState) {
-      return (
-        <DefaultRenderer
-          onNavigate={this.props.onNavigate}
-          navigationState={this.state.navigationState}
-        />
-      );
-    }
-    return null;
   }
+  return (
```

Comment:
Have you tried to return null if selectedKey != current scene? This way no double rendering should happen.